File: Apps/iOS/Features/Harmonica/HarmonicaPresenter.swift
import Foundation
import HarmonicaCore

struct HarmonicaPresenter {
    private static let noRecordingsText = "Sound is unavailable: the recordings are not in the app."
    private static let outputRefusedText = "Sound is unavailable: the audio output would not start."
    private static let overbendAtRestLabel = "overbend ↑"
    private static let overblowLabel = "overblow ↑"
    private static let overdrawLabel = "overdraw ↑"
    private static let bendLabel = "bend ↓"
    private static let vibratoLabel = "vibrato →"
    private static let cupLabel = "cup"
    private static let playDemoLabel = "▶\u{FE0E}"
    private static let stopDemoLabel = "stop"
    private static let bendEffectLabel = "bend"
    private static let overblowEffectLabel = "overblow"
    private static let overdrawEffectLabel = "overdraw"

    private let locale: Locale
    private let tunes: [HarmonicaViewState.Tune]
    private let holeLabels: [String]

    // MARK: - Public

    init(locale: Locale, tunes: [Score]) {
        self.locale = locale
        self.tunes = tunes.enumerated().map { HarmonicaViewState.Tune(id: $0.offset, name: $0.element.name) }
        holeLabels = Hole.allCases.map { $0.number.formatted(.number.locale(locale)) }
    }

    func present(_ harmonica: Harmonica, settings: PlayerSettings, playingAScore: Bool) -> HarmonicaViewState {
        .ready(
            HarmonicaViewState.Playable(
                holes: holes(of: harmonica),
                key: keyState(harmonica.key),
                fingerMarks: fingerMarksState(harmonica.style),
                cup: settings.isCuppingEnabled ? Self.cupState(harmonica.cup) : nil,
                demo: demoState(playingAScore),
                toneShaping: toneShaping(harmonica),
                square: HarmonicaViewState.Square(placement: settings.squarePlacement, size: settings.squareSize)
            )
        )
    }

    func presentSoundUnavailable(because failure: AudioEngineError) -> HarmonicaViewState {
        .soundUnavailable(Self.reason(for: failure))
    }

    // MARK: - Private

    private static func reason(for failure: AudioEngineError) -> String {
        switch failure {
        case .noRecordings: noRecordingsText
        case .recordingUnreadable(let name): "Sound is unavailable: the recording \(name) could not be read."
        case .outputRefused: outputRefusedText
        }
    }

    private func holes(of harmonica: Harmonica) -> [HarmonicaViewState.Hole] {
        zip(Hole.allCases, holeLabels).map { hole, label in
            holeState(hole, label: label, sounding: harmonica.sounding[hole])
        }
    }

    private func holeState(_ hole: Hole, label: String, sounding: SoundingReed?) -> HarmonicaViewState.Hole {
        HarmonicaViewState.Hole(
            id: hole.number,
            label: label,
            note: sounding.map { name(of: $0.pitch) } ?? "",
            effect: sounding.map(effect(shaping:)) ?? "",
            lit: sounding.map { Self.half(litBy: $0.breath) }
        )
    }

    private static func half(litBy breath: Breath) -> HarmonicaViewState.LitHalf {
        switch breath {
        case .blow: .top
        case .draw: .bottom
        }
    }

    private func effect(shaping reed: SoundingReed) -> String {
        guard reed.isShifted else { return "" }

        return "(\(name(of: reed.unbent)) \(effectName(shaping: reed)))"
    }

    private func effectName(shaping reed: SoundingReed) -> String {
        guard reed.isOverbent else { return Self.bendEffectLabel }

        return reed.breath == .blow ? Self.overblowEffectLabel : Self.overdrawEffectLabel
    }

    private func name(of note: MIDINote) -> String {
        MIDINote.namesAboveC[note.semitonesAboveC] + note.octave.formatted(.number.locale(locale))
    }

    private func toneShaping(_ harmonica: Harmonica) -> HarmonicaViewState.ToneShaping {
        HarmonicaViewState.ToneShaping(
            overbendLabel: Self.overbendLabel(breathing: harmonica.breath),
            bendLabel: Self.bendLabel,
            vibratoLabel: Self.vibratoLabel,
            isOverbendAvailable: harmonica.canOverbend,
            isBendAvailable: harmonica.canBend
        )
    }

    private static func overbendLabel(breathing breath: Breath?) -> String {
        switch breath {
        case .blow: overblowLabel
        case .draw: overdrawLabel
        case nil: overbendAtRestLabel
        }
    }

    private static func cupState(_ cup: CupDepth) -> HarmonicaViewState.Cup {
        HarmonicaViewState.Cup(label: cupLabel, closed: cup.fraction)
    }

    private func fingerMarksState(_ style: PlayingStyle) -> HarmonicaViewState.FingerMarks {
        HarmonicaViewState.FingerMarks(
            isDrawnAtTheContactWidth: style.coversTheContactWidth,
            drawsOnlyTheDecidingFinger: style.takesTheTopmostFingerOnly
        )
    }

    private func demoState(_ isPlaying: Bool) -> HarmonicaViewState.Demo {
        HarmonicaViewState.Demo(
            label: isPlaying ? Self.stopDemoLabel : Self.playDemoLabel,
            tunes: tunes,
            isPlaying: isPlaying
        )
    }

    private func keyState(_ key: HarmonicaKey) -> HarmonicaViewState.Key {
        HarmonicaViewState.Key(
            label: label(for: key),
            position: Double(key.sliderPosition),
            highestPosition: Double(HarmonicaKey.highestSliderPosition)
        )
    }

    private func label(for key: HarmonicaKey) -> String {
        switch key {
        case .g: "G"
        case .aFlat: "A♭"
        case .a: "A"
        case .bFlat: "B♭"
        case .b: "B"
        case .c: "C"
        case .dFlat: "D♭"
        case .d: "D"
        case .eFlat: "E♭"
        case .e: "E"
        case .f: "F"
        case .fSharp: "F♯"
        }
    }
}

File: Web/src/features/harmonica/harmonicaPresenter.ts
import type { AudioEngineError } from '../../core/audioEngine.js'
import type { SoundingHoleDTO, HarmonicaDTO } from '../../core/harmonicaDTO.js'
import { holesWideOfTheContact } from '../../core/harmonicaDTO.js'
import type { TuneDTO } from '../../core/tuneDTO.js'
import { keyNames, nameOf } from './noteNames.js'
import type {
    DemoViewState,
    FingerMarksViewState,
    FingerWidthViewState,
    HarmonicaViewState,
    HoleViewState,
    KeyViewState,
    LitHalf,
    NoteCountViewState,
    NotesPerFingerViewState,
    PlayingStyleChoice,
    PlayingStyleViewState,
    ShapingPadViewState,
    TuneViewState
} from './harmonicaViewState.js'
import { playingStyleChoices } from './harmonicaViewState.js'

const noRecordingsText = 'Sound is unavailable: the recordings are not on the page.'
const outputRefusedText = 'Sound is unavailable: the audio output would not start.'
const pitchAtRestLabel = 'bend · overbend ↑'
const bendLabel = 'bend ↑'
const overblowLabel = 'overblow ↑'
const overdrawLabel = 'overdraw ↑'
const vibratoLabel = 'vibrato →'
const playDemoLabel = '▶\uFE0E'
const stopDemoLabel = 'stop'
const bendEffectLabel = 'bend'
const overblowEffectLabel = 'overblow'
const overdrawEffectLabel = 'overdraw'
const notesByPressureLabel = 'by pressure'

const everyHole: readonly number[] = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
const noteCountChoices: readonly NoteCountViewState[] =
    [1, 2, 3, 4].map(count => ({ count, name: count === 1 ? '1 note' : `${count} notes` }))
const notesByPressureChoices: readonly NoteCountViewState[] =
    [{ count: holesWideOfTheContact, name: notesByPressureLabel }]

export class HarmonicaPresenter {
    private readonly tunes: readonly TuneViewState[]
    private readonly styleChoices: readonly PlayingStyleChoice[]

    constructor(tunes: readonly TuneDTO[], offersNotesByPressure: boolean) {
        this.tunes = tunes.map((tune, index) => ({ id: index, name: tune.name }))
        this.styleChoices = offersNotesByPressure
            ? playingStyleChoices
            : playingStyleChoices.filter(choice => choice !== 'oneFingerNotesByPressure')
    }

    present(harmonica: HarmonicaDTO, playingAScore: boolean, shapingPadScale: number): HarmonicaViewState {
        const style = choiceOf(harmonica)
        return {
            kind: 'ready',
            playable: {
                holes: holeStates(harmonica),
                key: keyState(harmonica.keyPosition),
                style: styleState(style, this.styleChoices),
                notesPerFinger: notesPerFingerState(style, harmonica.mouthHolesWide),
                fingerMarks: fingerMarksState(style, harmonica.mouthHolesWide),
                demo: this.demoState(playingAScore),
                shapingPad: shapingPadState(harmonica, shapingPadScale)
            }
        }
    }

    presentSoundUnavailable(because: AudioEngineError): HarmonicaViewState {
        return { kind: 'soundUnavailable', text: reasonFor(because) }
    }

    private demoState(isPlaying: boolean): DemoViewState {
        return {
            label: isPlaying ? stopDemoLabel : playDemoLabel,
            tunes: this.tunes,
            isPlaying
        }
    }
}

function reasonFor(failure: AudioEngineError): string {
    switch (failure.kind) {
        case 'noRecordings': return noRecordingsText
        case 'recordingUnreadable': return `Sound is unavailable: the recording ${failure.recording} could not be read.`
        case 'outputRefused': return outputRefusedText
    }
}

function holeStates(harmonica: HarmonicaDTO): HoleViewState[] {
    return everyHole.map(hole => holeState(hole, harmonica.sounding.find(each => each.hole === hole)))
}

function holeState(hole: number, sounding: SoundingHoleDTO | undefined): HoleViewState {
    return {
        id: hole,
        label: String(hole),
        note: sounding === undefined ? '' : nameOf(sounding.pitch),
        effect: sounding === undefined ? '' : effectOf(sounding),
        lit: sounding === undefined ? null : halfLitBy(sounding.breath)
    }
}

function halfLitBy(breath: 'blow' | 'draw'): LitHalf {
    return breath === 'blow' ? 'top' : 'bottom'
}

function effectOf(reed: SoundingHoleDTO): string {
    if (!reed.isShifted) return ''

    return `(${nameOf(reed.unbent)} ${effectNameOf(reed)})`
}

function effectNameOf(reed: SoundingHoleDTO): string {
    if (!reed.isOverbent) return bendEffectLabel

    return reed.breath === 'blow' ? overblowEffectLabel : overdrawEffectLabel
}

function shapingPadState(harmonica: HarmonicaDTO, scale: number): ShapingPadViewState {
    return {
        pitchLabel: pitchLabelFor(harmonica),
        vibratoLabel,
        isPitchShapingAvailable: harmonica.canBend || harmonica.canOverbend,
        scale
    }
}

function pitchLabelFor(harmonica: HarmonicaDTO): string {
    if (harmonica.canBend) return bendLabel
    if (!harmonica.canOverbend || harmonica.breath === null) return pitchAtRestLabel

    return harmonica.breath === 'blow' ? overblowLabel : overdrawLabel
}

function styleState(chosen: PlayingStyleChoice, offered: readonly PlayingStyleChoice[]): PlayingStyleViewState {
    return {
        chosen,
        label: nameOfChoice(chosen),
        choices: offered.map(each => ({ id: each, name: nameOfChoice(each) }))
    }
}

function choiceOf(harmonica: HarmonicaDTO): PlayingStyleChoice {
    switch (harmonica.style) {
        case 'severalFingersOneNote': return 'severalFingersOneNote'
        case 'oneFingerSeveralNotes': return oneFingerChoiceOf(harmonica.mouthHolesWide)
        default: return 'severalFingersSeveralNotes'
    }
}

function oneFingerChoiceOf(holesWide: number): PlayingStyleChoice {
    return holesWide === holesWideOfTheContact ? 'oneFingerNotesByPressure' : 'oneFingerSeveralNotes'
}

function nameOfChoice(choice: PlayingStyleChoice): string {
    switch (choice) {
        case 'severalFingersSeveralNotes': return 'Many fingers, many notes'
        case 'severalFingersOneNote': return 'Many fingers, one note'
        case 'oneFingerSeveralNotes': return 'One finger, many notes'
        case 'oneFingerNotesByPressure': return 'One finger, notes by pressure'
    }
}

function notesPerFingerState(style: PlayingStyleChoice, holesWide: number): NotesPerFingerViewState {
    switch (style) {
        case 'severalFingersOneNote':
            return { chosen: 1, choices: noteCountChoices, isAvailable: false }
        case 'oneFingerNotesByPressure':
            return { chosen: holesWideOfTheContact, choices: notesByPressureChoices, isAvailable: false }
        default:
            return { chosen: holesWide, choices: noteCountChoices, isAvailable: true }
    }
}

function fingerMarksState(style: PlayingStyleChoice, holesWide: number): FingerMarksViewState {
    return {
        width: fingerWidthState(style, holesWide),
        drawsOnlyTheDecidingFinger: style === 'oneFingerSeveralNotes' || style === 'oneFingerNotesByPressure'
    }
}

function fingerWidthState(style: PlayingStyleChoice, holesWide: number): FingerWidthViewState {
    switch (style) {
        case 'severalFingersOneNote': return { kind: 'resting' }
        case 'oneFingerNotesByPressure': return { kind: 'pressed' }
        default: return { kind: 'holesWide', holes: holesWide }
    }
}

function keyState(position: number): KeyViewState {
    return {
        label: keyNames[position] ?? '',
        position,
        highestPosition: keyNames.length - 1
    }
}
